### PR TITLE
[SYCL][Clang] Fix address space for virtual table support

### PR DESCRIPTION
**Pull request by elizabethandrews**
For SPIR, virtual table elements should be emitted in the default address space while global variables should be emitted in address space 1.

Also, only emit virtual functions marked with [[intel::device_indirectly_callable]]. This attribute will be replaced by compile time properties (for this check) once the spec for virtual function extension is finalized.

Patch by: Michael Rice (michael.p.rice@intel.com) and Elizabeth Andrews

### CODE REVIEW

File: clang/test/CodeGenSYCL/sycl-device-only-virtual-functions.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
+// Purpose of this test is to ensure that in SYCL device code mode, we are only
+// emitting those virtual functions into vtable, which are marked with
+// [[intel::device_indirectly_callable]] attribute.
```
Comment (AlexeySachkov) [inline clang/test/CodeGenSYCL/sycl-device-only-virtual-functions.cpp:3]: Implementation also allows emission of functions which are marked with `sycl_device` attribute only, if I understand correctly. That attribute will likely be the only thing we check to decide if we need to emit a virtual function into device code going forward. Therefore, I think that it is important to have it checked as well, i.e. in this test we should replace a couple of `[[intel::device_indirectly_callable]]` occurrences with `SYCL_EXTERNAL` and that should not change the resulting LLVM IR

Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/sycl-device-only-virtual-functions.cpp]: >>That attribute will likely be the only thing we check to decide if we need to emit a virtual function into device code going forward. 
Actually do you want to remove support for  [[intel::device_indirectly_callable]] then ? I can do it here and then make changes downstream as well.

Comment (AlexeySachkov) [reply clang/test/CodeGenSYCL/sycl-device-only-virtual-functions.cpp]: > Actually do you want to remove support for `[[intel::device_indirectly_callable]]` then ? I can do it here and then make changes downstream as well.

We may have some internal customers relying on behavior of this attribute, so I would keep it for now. I will work on its removal alongside with the extension implementation

**Review by AlexeySachkov — APPROVED** [review]

**Comment by elizabethandrews** [comment]
Hmm looks like we have some failing lit tests. Not sure why they didn't fail locally. Taking a look now.

**Review by premanandrao — APPROVED** [review]
LGTM

**Review by smanna12 — APPROVED** [review]

**Comment by elizabethandrews** [comment]
@intel/llvm-gatekeepers this is ready for merge.

**Comment by ldrumm** [comment]
I'm confused about the authorship here. Is this a case where one of these commits is by you, Elizabeth, and one is Michael?

**Comment by elizabethandrews** [comment]
ping @intel/llvm-gatekeepers do you need more clarification on authors? I think you can set either Mike or I work as authors here.

**Comment by sarnex** [comment]
I added Co-authored-by to the commit message for Mike, that should capture everything.

**Comment by ldrumm** [comment]
Thanks for the clarification. That makes sense. I wanted to avoid an issue we've been having with misattributed cherry-picks.

**Pull request merged**